### PR TITLE
[WIP] tiled synthesis simulation

### PR DESCRIPTION
**Pull request by Starnite**

### CODE REVIEW

**Comment by ericmjl** [comment]
Good start! Might need you to walk me through your code in person. Three things that I'm currently checking for - the tiles are overlapping, and that they're in DNA rather than amino acids, and that both the forward and reverse strands are available. 

I also found some papers that might give some inspiration:

- [Multiplex pairwise assembly](http://nar.oxfordjournals.org/content/early/2015/11/06/nar.gkv1177.long)
- [Dial-out PCR](https://www.ncbi.nlm.nih.gov/pubmed/22886093)

Give them a read-through, and let's talk about them in the lab (whenever you're in).

**Comment by Starnite** [comment]
Do you remember what search criteria we used when going through the database? I’m searching for 2011-2016, human H3N2 type A PB2, but I’m getting a lot more results than I downloaded before break.

**Comment by ericmjl** [comment]
I don't remember the search parameters, but you could try removing duplicate sequences (it's one of the additional options available). Good reminder this time round to record the search params down too. :smile:

**Comment by ericmjl** [comment]
Great work!

Here's another consideration - I think there's many more unique nucleotide sequences than unique protein sequences. Makes sense, because of the degenerate 3rd codon. If you were to use a reverse translation (AA -> DNA) of the protein sequences vs. straight-up using known nucleotide sequences, what's the difference in the number of unique tiles that have to be synthesized?

**Comment by Starnite** [comment]
I originally looked for unique nucleotide sequences in the database (there are nearly 3000), but my thinking was that for the most part, the polymorphisms wouldn’t really matter unless it resulted in a translation to a different amino acid. If I did reverse translation, I could either do every possible codon combination, or the most common codon for influenza.

**Comment by ericmjl** [comment]
True, nucleotide degeneracy in the 3rd codon position might not matter for protein expression, but from an on-chip DNA synthesis standpoint, it very likely will matter. Each degenerate sequence is a new strand that needs to be made, and there's an upper-limit (not sure how much right now, but assumed to be on the order of 10K-100K) of the number of unique nucleotide strands that can be synthesized on one chip.

**Comment by Starnite** [comment]
I might be misunderstanding something here – wouldn’t there be far too many possibilities if we took into account every degeneracy for every amino acid in the protein?

I will be back tomorrow afternoon – snow got me stuck in NY for an extra day. See you soon!

**Comment by ericmjl** [comment]
You're right!! I was hoping you'd highlight this independent of myself
telling you that point :D.

What would be the engineering implications of this point? If the goal is
"make protein sequence (in fragments) while minimizing number of nucleotide
fragments required", how would you go about designing the oligos to be
synthesized on-chip?

**Comment by Starnite** [comment]
The two articles I mentioned today:

http://www.nature.com/nature/journal/v432/n7020/full/nature03151.html
http://www.nature.com/nbt/journal/v28/n12/full/nbt.1716.html

See you tomorrow!

**Comment by ericmjl** [comment]
Progress looking good. I think a good next step is to refactor the code into a Python module that you can import elsewhere, and write a few small-scale test cases that can be used to check for correctness in implementation.

File: experiments/20170207-tiled-synthesis-computation/Goals-February.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+## Goals for the month of February
```
Comment (ericmjl) [inline experiments/20170207-tiled-synthesis-computation/Goals-February.md]: @Starnite, separate PR for goals please! :smile:

**Review by ericmjl — CHANGES_REQUESTED** [review]

**Comment by ericmjl** [comment]
@Starnite: can you demo your code for me this Thursday, say in the morning? It's so that I can get a better feel for it.

**Comment by Starnite** [comment]
Sure! Would 10am work for you?

**Comment by ericmjl** [comment]
Let's do that. What's your calendar email address? MIT or Gmail?

**Comment by Starnite** [comment]
Gmail – vivianzhong06@gmail.com

**Comment by Starnite** [comment]
@ericmjl For 30-bp overlaps and threshold = 1, and testing with 5 sequences to start with, I have 11 re-constructed sequences that are present in the original pool, out of 1896 total.

**Comment by ericmjl** [comment]
@Starnite Hmm... a bit confused here, how do we start with 5 sequences and end up with 11 reconstructed out of the original 5?

**Comment by Starnite** [comment]
Many tiles are identical, so there are duplicates of certain sequences – so they’re different paths, but because the nodes are the same, the sequence is the same.

**Comment by ericmjl** [comment]
Got it. This is interesting, it means that the pool of non-original
sequences from the tiles is quite large, under the conditions where one
mismatch per tile is allowed. This is good information; it'll elevate the
importance of PacBio sequencing + barcoding in rationally arraying the
protein variants.

Can you do me a favour and test the code on influenza neuraminidase? Use
global H3N2 neuraminidase sequences from the IRD, starting from year 2000
to present. Under the assumption of only perfect matches, how many
originals out of the space of total possibles are there?

**Comment by Starnite** [comment]
Sure, will do!

**Comment by Starnite** [comment]
Hi Eric,

With tile lengths of 210 (30bp overlap on each end included), and testing with 10 sequences, I get 108 original out of 61666 total.

**Comment by ericmjl** [comment]
Great stuff! Now give it a shot on all sequences. If you want to offload the computation to the GPU workstation remotely, go for it!

Btw, if you're using a for-loop in your code, check out the Python package `tqdm`. It'll give you an estimate for the total loop time needed early on, which you can use to determine whether to offload or not.

Another btw: I think there's a matrix interpretation of path-finding between two nodes, even for directed acyclic graphs (which is what you're constructing). That might be really useful for speeding up computation. Check out this StackExchange question: http://math.stackexchange.com/questions/1890620/finding-path-lengths-by-the-power-of-adjacency-matrix-of-an-undirected-graph.

**Comment by Starnite** [comment]
Thanks for the tips Eric! Is there a way to run the program in the GPU server without having to re-install all the packages?

**Comment by ericmjl** [comment]
I think you can export your conda environment as a YAML file, and then use
the `conda create` command to re-create your compute environment on the GPU
workstation using the YAML file.

Check out the conda package manager docs:
https://conda.io/docs/using/envs.html

**Comment by Starnite** [comment]
Hmm…I’m currently trying to use the ‘scp’ command to copy the yml file to the server, but I get “permission denied”.

**Comment by ericmjl** [comment]
Hmm.... where are you copying to?

**Comment by Starnite** [comment]
vivzhong@ubuntu-gpu.myddns.me

**Comment by ericmjl** [comment]
Let me try resetting your pw.

**Comment by ericmjl** [comment]
Try again. Your pw is vivianzhong@mit.edu.

To change your pw, after you login, use "passwd vivzhong", and quickly
change it.

Let's see if that fixes the permissions error.

Finally, just checking, you aren't scp-ing to the root directory, but your
home directory right? The destination should start with "~/".

**Comment by Starnite** [comment]
Thanks Eric! So I managed to scp the environment.yml file into the home directory, but it seems like I’d still need to do some installing with sudo<http://datascienceguide.github.io/how-to-install-the-python-data-science-stack-on-a-remote-server> and pip in order to use the conda command, and I don’t have sudo privileges.

**Comment by ericmjl** [comment]
Great progress! I think you can use the Anaconda distribution of Python to
get around needing to use sudo privileges.

**Comment by Starnite** [comment]
I have everything working! But the code as is definitely takes far too long to run, so I’ll work on optimizing it. With 50 starting sequences, we get 4337 original and 290651837 new sequences.

**Comment by ericmjl** [comment]
Good stuff! I notice that your code is particularly useful for estimating the proportion of "desired" sequences that can show up. You might want to do a Monte Carlo walk on the graph, as an alternative in case optimization takes a long time.

**Comment by Starnite** [comment]
Hey Eric,

Sorry for falling off the grid! I tried to do what I tried to do what I thought would be a simple simplification, but it ended up making the runtime even longer.

My initial method of adding edges was to simply evaluate every pairing of tiles in the digraph as determined by itertools.product. I figured that this was pretty unnecessary, and changed the code so that it only compared two tiles if the second came right after the first in the DNA sequence. This actually resulted in more edges being added, and thus exponentially more paths to search and exponentially longer runtime.

For example, with 2 sequences, 8 tiles each (for the neuraminidase):

First method:
num nodes: 16
num edges: 19
original: 2 new: 5
Second method:
num nodes: 16
num edges: 24
original: 2 new: 62

The case where every overlap is perfect should result in 28 edges (2 edges per node in a sequence * 7 nodes in a sequence (excluding the last one) * 2 sequences), so I’m inclined to think that the second method is correct. I’m not sure why the first method lost a few edges, though – perhaps itertools.product isn’t getting everything?

**Comment by Starnite** [comment]
Addendum: I checked the runtime using the second method for a few different numbers of sequences, and extrapolating from a fitted exponential curve, the code as is would take 68 days to sort through ~2000 sequences. So I’ll definitely look into random walk MC.

**Comment by ericmjl** [comment]
Hmm... It looks like it'll be worth us meeting up to look through some code. I have a suspicion that you'll need to incorporate more test cases in test suite, but let's chat in person. Got a time that works next week?

**Comment by Starnite** [comment]
Hi Eric,

Would Thursday 2:30 work for you? I unfortunately have an exam on Wednesday.

I know you mentioned earlier this semester that you wanted to record me carrying out the 96-well polymerase assay; do you have an approximate date in mind for that?

**Comment by ericmjl** [comment]
@Starnite Thursday's good; I'll be at the Sloan School of Mgmt all day though (there's a 1-day conference going on there), will it be okay with you to walk over there? If not, we can always do a Hangout call after hours too.

**Comment by ericmjl** [comment]
As for videographing the 96-well plate assay, let me work out some other experimental details; there's still one transfection experiment I'll need your help with as well, but I need to make sure the protocol we have is good to go.

**Comment by Starnite** [comment]
Hi Eric,

My 6.036 project deadline just got pushed to Monday, so I’d be happy to make the trip. :) What time works for you?

**Comment by ericmjl** [comment]
@Starnite I took a look at the agenda for the day, it looks like everything ends at 4 pm. Do you think that's a good time for you? I originally thought 2:30 pm, but as it turns out there's only a 15 min. break there -_-", but I think your stuff is worth spending a good hour (or more) on.

**Comment by Starnite** [comment]
Yup, 4 is good!





On Wed, Apr 5, 2017 at 10:49 PM -0400, "Eric Ma" <notifications@github.com<mailto:notifications@github.com>> wrote:


@Starnite<https://github.com/Starnite> I took a look at the agenda for the day, it looks like everything ends at 4 pm. Do you think that's a good time for you? I originally thought 2:30 pm, but as it turns out there's only a 15 min. break there -_-", but I think your stuff is worth spending a good hour (or more) on.

**Comment by ericmjl** [comment]
ping @Starnite: as promised yesterday, let's merge this and start a new PR to continue the conversation!

**Review by ericmjl — APPROVED** [review]

**Pull request merged**